### ISSUE TITLE
[HOLD for payment 2025-02-26] [$250] Expense report -"Required" Error Not Triggered for Auto-Filled Report Field on Second Expense Submission

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#47251]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v9.0.19-0
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**Email or phone of affected tester (no customers):** abebemiherat@gmail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Go to FAB > Workspace.
2. Navigate to Workspace Settings > More Features > Enable Report Field.
3. In Report Field, add a field, name it, choose a type, and save.
4. Go to Workspace Chat > Click the plus sign > Submit Expense > Add amount, merchant, and submit.
5. In Expense report, fill in the Report Field > Notice that the error triggers > Go back to the workspace chat, click Submit, and select Pay Elsewhere.
6. Submit a second expense by repeating Step 4 > Go to Expense report > Notice that the Report Field is auto-filled, but the error does trigger.

## Expected Result:
The "Required" error message should not appear if the Report Field in the second expense is auto-filled with information from the first expense.

## Actual Result:
The "Required" error message does not trigger for the Report Field on the second submitted expense when auto-filled.

## Workaround:
Unknown

## Platforms:

- [x] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/69cc30dc-7e05-47f5-a160-6bc73c415c73

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0185bccf1e3226865f</li>
        <li>Upwork Job ID: 1823114245300473767</li>
        <li>Last Price Increase: 2024-09-02</li>
<li>Automatic offers: </li>
<ul>
<li>dukenv0307 | Reviewer | 103834642</li>
<li>nkdengineer | Contributor | 103834644</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @VictoriaExpensify</details>

**Comment by IuliiaHerets** [comment]
We think that this bug might be related to #vip-vsb

**Comment by VictoriaExpensify** [comment]
Agree this is an issue and we should fix it.

**Comment by VictoriaExpensify** [comment]
Report Fields are part of Wave_Control, so I'm adding this issue to that project

**Comment by Beamanator** [comment]
Report Fields are behind a beta, so i'm demoting this one

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The "Required" error message does not trigger for the Report Field on the second submitted expense when auto-filled.

### What is the root cause of that problem?
When we create an expense report, the `fieldList` is assigned as the `fieldList` of the policy.

https://github.com/Expensify/App/blob/420c2a13cc176dca5adf8dcc8132ef5f8ff678f7/src/libs/ReportUtils.ts#L4285

Then because the field has default value is empty, a report violation is added here.
https://github.com/Expensify/App/blob/420c2a13cc176dca5adf8dcc8132ef5f8ff678f7/src/libs/actions/IOU.ts#L480-L486

After BE returns the data, this field is auto-filled but the report violation isn't cleared then the error still appears
### What changes do you think we should make in order to solve the problem?
When we create an expense report here, we should auto-fill the report field from the last paid expense report of this policy expense chat (we can confirm the logic from BE). 

https://github.com/Expensify/App/blob/420c2a13cc176dca5adf8dcc8132ef5f8ff678f7/src/libs/ReportUtils.ts#L4285

Then the violation will not appear here since the value already exists
https://github.com/Expensify/App/blob/420c2a13cc176dca5adf8dcc8132ef5f8ff678f7/src/libs/actions/IOU.ts#L480-L486

### What alternative solutions did you explore? (Optional)

If the violation is` fieldRequired` and the value exists we should return an empty string here

https://github.com/Expensify/App/blob/420c2a13cc176dca5adf8dcc8132ef5f8ff678f7/src/libs/ReportUtils.ts#L7471

and should not display a red dot here
https://github.com/Expensify/App/blob/420c2a13cc176dca5adf8dcc8132ef5f8ff678f7/src/components/ReportActionItem/MoneyReportView.tsx#L122

**Comment by nkdengineer** [comment]
@dukenv0307 Update my [proposal](https://github.com/Expensify/App/issues/47251#issuecomment-2285295748) with implementation detail. We can confirm with the internal team about this logic and the BE logic.

**Comment by puneetlath** [comment]
It feels to me like this is working as expected. But confirming here: https://expensify.slack.com/archives/C06ML6X0W9L/p1723746405468429

**Comment by nkdengineer** [comment]
@puneetlath The problem here is I want to know how BE auto-fill the report field for an expense report when we create an expense. So I can do this in optimistic data to prevent the incorrect report violation in optimistic data.

**Comment by mountiny** [comment]
@nkdengineer we are still discussing this one. It seems like the report fields should be sticky between the reports but when I tried on OldDot, it was not sticky. Can you try as well what is the different behaviour between NewDot and OldDot in this flow? Thanks!

**Comment by nkdengineer** [comment]
@mountiny I tested in OldDot, it was sticky

https://github.com/user-attachments/assets/e2d9e0a2-6415-476f-85b2-cfba197181f2

**Comment by mountiny** [comment]
@nkdengineer what if you create a new workspace and report and set up the report fields in oldDot solely (sorry mobile today so cant test myself) - i think that was the flow where i got no stickiness

**Comment by nkdengineer** [comment]
> what if you create a new workspace and report and set up the report fields in oldDot solely (sorry mobile today so cant test myself) - i think that was the flow where i got no stickiness

@mountiny I tested this case and the new report is also auto-filled by previous paid expense report after we paid an expense report.

**Comment by puneetlath** [comment]
It sounds like everything is working correctly then, is that right?

**Comment by nkdengineer** [comment]
@puneetlath Yes BE is working correctly but we need to update the optimistic data in front-end to prevent incorrect violation is displayed.

> The problem here is I want to know how BE auto-fill the report field for an expense report when we create an expense. So I can do this in optimistic data to prevent the incorrect report violation in optimistic data.

Can you check my concern above?

cc @mountiny

**Comment by mountiny** [comment]
@nkdengineer Yes you are right, we need to update the optimistic case, althouhg I wonder if we have all the data for this available when user signs out and signs back in, can you check that and make a proposal for that case?

**Comment by mountiny** [comment]
Not overdue

**Comment by mountiny** [comment]
@nkdengineer are you able to look into this one further after my latest comment?

**Comment by nkdengineer** [comment]
I'm checking this by testing some cases with many report previews. Will give an update tomorrow.

**Comment by nkdengineer** [comment]
@mountiny I checked, and sometimes the last paid expense report doesn't exist when we logout and login again. Can we return the `fieldList` of the last paid expense report in the data of policyExpenseChat report like the `iouReportID` field that is the last unsettled expense report?

**Comment by mountiny** [comment]
Hmm I am not sure if this is really a feasible approach for us. Is there a way we could just avoid showing the optimistic violation rather and wait for the onyx data to come from the backend to only then show this violation @nkdengineer

**Comment by mountiny** [comment]
could we do both:
- detect we have the fieldList data we need to potentially show the optimistic violation and show it
- if we do not have the necessary data locally do not show the violation optimistically

**Comment by nkdengineer** [comment]
> detect we have the fieldList data we need to potentially show the optimistic violation and show it

@mountiny  I think it's hard to detect, for example, we cannot know if the paid expense report is the latest or not or cannot know if we have the paid or not because in `OpenApp` API, the data may not contain all `REPORTPREVIEW` actions or expense reports of the policy expense chat

**Comment by mountiny** [comment]
@nkdengineer Yeah this is not easy thing to resolve. I need to think about this more. What would you propose now to handle this gracefully without adding any more data from backend?

**Comment by VictoriaExpensify** [comment]
Not overdue

**Comment by nkdengineer** [comment]
> What would you propose now to handle this gracefully without adding any more data from backend?

@mountiny I propose to not add field violation to optimistic data here if we create a new expense report.

https://github.com/Expensify/App/blob/b9af3d1f7ab160f7d817102d081b0e6186ba2d02/src/libs/actions/IOU.ts#L888-L890

**Comment by mountiny** [comment]
@nkdengineer Are none of the reliably computable optimistically? I agree it might be cleaner to not show this optimistically as they can be complex and wait for backend

**Comment by nkdengineer** [comment]
>  Are none of the reliably computable optimistically? I agree it might be cleaner to not show this optimistically as they can be complex and wait for backend

@mountiny  Yes, only when we edit the report field we can computable the optimistic data correctly because in this case, we can know exactly the value of report field.

**Comment by puneetlath** [comment]
Could we get it when opening the report instead of when opening the app?

**Comment by nkdengineer** [comment]
> Could we get it when opening the report instead of when opening the app?

We can create the money request via FAB without opening the report.

**Comment by mountiny** [comment]
> We can create the money request via FAB without opening the report.

+1

**Comment by puneetlath** [comment]
Ahh ok, sorry I didn't correctly understand. Would it be possible to create them optimistically _if_ we have the data, but don't create them optimistically if we don't? 

Otherwise, what you suggest makes sense to me.

**Comment by VictoriaExpensify** [comment]
Not overdue

**Comment by nkdengineer** [comment]
> detect we have the fieldList data we need to potentially show the optimistic violation and show it

> @mountiny I think it's hard to detect, for example, we cannot know if the paid expense report is the latest or not or cannot know if we have the paid or not because in OpenApp API, the data may not contain all REPORTPREVIEW actions or expense reports of the policy expense chat

> Ahh ok, sorry I didn't correctly understand. Would it be possible to create them optimistically if we have the data, but don't create them optimistically if we don't?

@puneetlath We discussed this above, and then we go with the final solution.

**Comment by puneetlath** [comment]
Ok got it! Sounds good to me.

**Comment by thienlnam** [comment]
Catching up in this thread, but it looks like we just decided to remove optimistic report field violations.

When report fields get updated, there should be a onyx update sent out to update the field list. So the field list should be up to date if you're online. So why can't we just not show the violation when the field is auto-filled?

**Comment by mountiny** [comment]
From what I understood is that if you sign out and sign in, we have no way to know if the field should be auto-filled/ what it should be autofilled with. Instead of adding more data (in this case it might be quite complex) to the OpenApp response, we thought it will be cleaner to just not show this optimistically if we cannot guarantee we know the outcome always and wait for the server response.

Please let us know if there is something we missed, but I think that on fresh sign in, we cannot know what field to autofill

**Comment by thienlnam** [comment]
On a fresh sign in, the policies should be fetched right? And the fieldList is part of the policy onyx key so it should also be loaded unless it gets loaded with like a condensed amount of data

**Comment by thienlnam** [comment]
But also, that edge case feels like like it is not as problematic as not showing report violations at all optimistically, do you agree with that?

**Comment by mountiny** [comment]
When we load all the policies, they are formed [here](https://github.com/Expensify/Web-Expensify/blob/4ae6352a742799888f839c562ed8e080ffcc59bd/lib/AppInit.php#L87-L118) and there is no `fieldList` there, I think we omit the large lists like tags or categories for policies that are not primary.

I am not sure, I think we are comparing a false positive error rarely and then no feedback rarely. That is once we load the policy and report with the reportFields we are good. I prefer the no feedback rather than false positive error

**Comment by thienlnam** [comment]
In that case, we can just return the fieldList there as well. That field isn't as large as tags or categories which is why we merged it as part of the policy key, and that should resolve this problem - does that sound good?

**Comment by mountiny** [comment]
We can try it, I did not want to add more load on the OpenApp call, but we can evaluate later

**Comment by nkdengineer** [comment]
@mountiny Is it ready for payment?

**Comment by mountiny** [comment]
I would still first try to handle the backend change, I will try to look into it this week

**Comment by nkdengineer** [comment]
@mountiny friendly bump

**Comment by mountiny** [comment]
Still on my list

**Comment by nkdengineer** [comment]
@mountiny Any update here?

**Comment by mountiny** [comment]
Haven't got around to this one yet, handling app deploy this week

**Comment by mountiny** [comment]
@nkdengineer can you add the checklist please?

**Comment by mountiny** [comment]
I dont think we need a specific test for this

**Comment by VictoriaExpensify** [comment]
**Payment summary:**
C: @nkdengineer paid $250 via Upwork
C+: @dukenv0307 paid $250 via Upwork

Thanks for getting this one finished up!
